include/kframework/gas.md

```diff
@@ -39,12 +42,223 @@ module GAS
     rule minGas(I1:Int, I2:Int) => minInt(I1, I2)
     rule gas2Int(G:Int) => G
 endmodule
+```
+
+KEVM Infinite Gas
+=================
+
+Here we use the construct `#gas` to represent positive infinity, while tracking the gas formula through execution.
+This allows (i) computing final gas used, and (ii) never stopping because of out-of-gas.
+Note that the argument to `#gas(_)` is just metadata tracking the current gas usage, and is not meant to be compared to other values.
+As such, any `#gas(G)` and `#gas(G')` are the _same_ positive infinite, regardless of the values `G` and `G'`.
+In particular, this means that `#gas(_) <Gas #gas(_) => false`, and `#gas(_) <=Gas #gas(_) => true`, regardless of the values contained in the `#gas(_)`.
+
+```k
+module INFINITE-GAS
+    imports GAS
+
+    syntax Gas ::= #gas(Int) [klabel(infGas), symbol, smtlib(infGas)]
+
+    rule #gas(G) +Gas G'       => #gas(G +Int G')
+    rule #gas(G) -Gas G'       => #gas(G -Int G')
+    rule #gas(G) *Gas G'       => #gas(G *Int G')
+    rule #gas(G) /Gas G'       => #gas(G /Int G')  requires G' =/=Int 0
+    rule      G  +Gas #gas(G') => #gas(G +Int G')
+    rule      G  -Gas #gas(G') => #gas(G -Int G')
+    rule      G  *Gas #gas(G') => #gas(G *Int G')
+    rule      G  /Gas #gas(G') => #gas(G /Int G')  requires G' =/=Int 0
+    rule #gas(G) +Gas #gas(G') => #gas(G +Int G')
+    rule #gas(G) -Gas #gas(G') => #gas(G -Int G')
+    rule #gas(G) *Gas #gas(G') => #gas(G *Int G')
+    rule #gas(G) /Gas #gas(G') => #gas(G /Int G')  requires G' =/=Int 0
+
+    rule _:Int    <Gas #gas(_) => true
+    rule #gas(_)  <Gas _       => false [simplification]
+    rule #gas(_) <=Gas _:Int   => false
+    rule _       <=Gas #gas(_) => true  [simplification]
+
+    rule minGas(#gas(G), #gas(G')) => #gas(minInt(G, G'))
+    rule minGas(G:Int  , #gas(G')) => #gas(minInt(G, G'))
+    rule minGas(#gas(G), G':Int)   => #gas(minInt(G, G'))
+
+    rule gas2Int(#gas(G)) => G
+
+    rule #allBut64th(#gas(G)) => #gas(#allBut64th(G))
+    rule Cgascap(SCHED, #gas(GCAP), #gas(GAVAIL), GEXTRA) => #gas(Cgascap(SCHED, GCAP, GAVAIL, GEXTRA)) [simplification]
+    rule Cgascap(SCHED, #gas(GCAP), GAVAIL:Int, GEXTRA)   => #gas(Cgascap(SCHED, GCAP, GAVAIL, GEXTRA)) [simplification]
+    rule Cgascap(SCHED, GCAP:Int, #gas(GAVAIL), GEXTRA)   => #gas(Cgascap(SCHED, GCAP, GAVAIL, GEXTRA)) [simplification]
+
+    rule #if B #then #gas(G) #else #gas(G') #fi => #gas(#if B #then G #else G' #fi) [simplification]
+endmodule
+```
+
+KEVM Gas Fees
+=============
+
+Here are some internal helper functions for calculating gas. Most of these functions are specified in the YellowPaper.
+
+```k
+module GAS-FEES
+    imports GAS-SYNTAX
+    imports SCHEDULE
+
+    // MOVED FROM evm.md:173`
+    syntax AccountCode ::= Bytes
+    
```

Comment:
This can probably go in `evm-types.md`, which is a base file defining the basic sorts needed for KEVM.